[ROCm] Adding rocmprim as a third_party library, minor update to eigen patch

deven-amd opened the pull request:
This PR does two things

1. Add rocmprim as a third_party library, add it to the bazel target(s) which will eventually need it (in PRs to TF operators to be filed soon)

2. Make a minor update to the eigen patch.
The updates are to fix a couple of ROCm/HIP specific errors in the Eigen code. We have already begun the process of upstreaming these changes to Eigen. So this patch update is expected to be temporary in nature.

This PR is pre-requisite to enabling ROCm support for some of TF Operators. Please exepdite it as much as possible.

Thanks

deven

-------------------------------------

@tatianashp , @whchung : just FYI

deven-amd committed:
update to pull in rocprim as a third_party library

deven-amd commented:
rebased to remove merge conflicts

whchung commented on tensorflow/workspace.bzl:
```diff
@@ -709,6 +709,17 @@ def tf_workspace(path_prefix = "", tf_repo_name = ""):
+        sha256 = "12adf5bf3641d73c92915f102b17951f978704551fdcb9ed7f6311ed299b1d80",
+        strip_prefix = "rocPRIM-eff7d0687baf57db2507a31663a3dea72eed9093",
+        urls = [
+            "https://mirror.bazel.build/github.com/ROCmSoftwarePlatform/rocPRIM/archive/eff7d0687baf57db2507a31663a3dea72eed9093.tar.gz",
```
the package of `rocPRIM` in use LGTM

whchung reviewed (dismissed):
LGTM. The PR introduces `rocPRIM` dependency, as well as HIP-specific patch to `Eigen`, which matches what's being used in AMD downstream fork.

gunan commented:
Adding @rmlarsen for the changes in eigen patch.
Please do not merge without his approval.

deven-amd commented:
* rebased to remove merge conflict
* also added a commit to fix a compile error that got recently introduced in the `--config=rocm` build

----

@rmlarsen 

For now, I am leaving in the commit for the eigen patch. We can drop that commit from this PR, once the corresponding Eigen PR has been merged + integrated into TF. 

There are a few other PRs that are built on top of this PR, so dropping the eigen commit from this PR (before the TF integration of the Eigen PR) would break the compile for this + all those PRs.

deven-amd commented:
@rmlarsen 

Thank you for getting the Eigen PR merged. I appreciate your help in expediting that process.

Please update this PR once the TF eigen pointer is updated as well. At that point I will rebase this PR and drop the commit for the eigen patch file update.

Thanks again

deven

deven-amd commented:
@rmlarsen 

gentle ping.  would you happen to have any insight on when the TF eigen pointer will move foward to include the eigen PR that was merged?

thanks

deven-amd commented:
@rmlarsen , thank you for updating the TF eigen pointer.

I have re-based this PR to drop the appends to the eigen patch file. 
The only change that remains is the minor change 
from :  `#if defined(EIGEN_CUDA_ARCH)`
to : `#if defined(EIGEN_CUDA_ARCH) || defined(EIGEN_HIP_DEVICE_COMPILE)`
My understanding is that you will upstream this change to Eigen, (if needed). Let me know if that is incorrect.

Please review and approve this PR. It is a pre-req for quite a few other PRs :)

thanks

deven-amd commented:
@gunan @rmlarsen 

gentle ping

please review and merge the PR. This PR is a gating item for quite a few other ROCm PRs.

thanks

deven-amd commented:
@gunan @rmlarsen

gentle ping

angerson reviewed (commented):
The build stuff looks fine to me, but still depends on @rmlarsen's confirmation of the eigen patch.

angersson commented:
This change was approved, but is failing internally because third_party dependency changes need special handling for conversion. I'm working on it.

angersson commented:
@deven-amd I'm running into trouble merging this due to differences in internal behavior for `is_rocm/cuda` and `is_rocm/cuda_configured`. Can you explain why you needed to make changes with the `cuda` calls? Favoring `is_rocm` where `is_cuda` is used (and similar for where `is_cuda_configured`) would probably resolve these issues.

deven-amd commented:
@angersson 

@chsigg had a similar query in another PR...please see the discussion here (https://github.com/tensorflow/tensorflow/pull/26753#discussion_r267022934) for details.

The summary of it is that, the change from `if_cuda / if_rocm` to `if_cuda_is_configured / if_rocm_is_configured` is needed in scenarios where there is a common target being added by by both (`rocm` and `cuda`) to the dependency list. 

Using `if_cuda / if_rocm` (which returns a select statement) in such scenarios results in a bazel error (duplicate depedency). 

Switching to `if_cuda_is_configured / if_rocm_is_configured` (which returns the `true` arg or empty list) resolves the bazel error.

Hope this clears things up.

angersson commented:
@chsigg Can you look at importing this, or find someone who can, please? Thanks!

chsigg reviewed (dismissed):
The PR looks ok, but I would like to eventually come up with a better solution to our if_*() vs if_*_is_configured(). Maybe adding an `if_gpu(if_cuda_or_rocm, otherwise)` is the right way to go.

deven-amd commented:
@chsigg   gentle ping

how close are we to getting this PR merged?

thanks

deven-amd commented:
the CI failures seem unrelated to the changes in this PR....not sure what is going on

ezhulenev commented:
This is a problem with a internal/external build setup mismatch. I'll try to merge it tomorrow, but probably I'll have to find someone who understands our build tooling better than me... might take few days.

ezhulenev commented:
I've merged non-build related changes, could you please sync it with master, I'll try to find someone who can help with build-related changes.

deven-amd commented:
@ezhulenev , thanks for merging the non-build commits.

I will rebase this branch onto master, and push it out later today. After the update, there will be only one commit left in this PR (build change for rocprim)

deven-amd commented:
@ezhulenev rebase done

gunan commented:
I am trying to merge this, but this is proving to be very very difficult.
I had to revert all the changes from "if_cuda" to "if_cuda_is_configured"
Next time, let's split such changes into two, where one only adds the new external dependency, and another where we start using it.

deven-amd commented:
@gunan thank you for getting this merge done

will keep your suggestions in mind, the next time we need to add a dependency.

deven-amd commented:
@gunan 

> I had to revert all the changes from "if_cuda" to "if_cuda_is_configured"

The change from `if_cuda` to `if_cuda_is_configured` was intentional and needed for the `--config=rocm` build to work. With that change reverted, the `--config=rocm` build is broken :(

How can I help you resolve this such that `--config=rocm` build is working again?

@chsigg, just FYI to keep you in the loop

thanks

deven

----------------

build errors I get when trying to do a `--config=rocm` build
```
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4029:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'bias_op_gpu'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4029:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'bias_op'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4029:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'libtfkernel_bias_op.so'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4069:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'softmax_op_gpu'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4069:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'softmax_op'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4069:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'libtfkernel_softmax_op.so'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4735:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'sparse_xent_op_gpu'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4735:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'sparse_xent_op'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4735:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'libtfkernel_sparse_xent_op.so'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:5227:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'multinomial_op_gpu'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:5227:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'multinomial_op'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:5227:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'libtfkernel_multinomial_op.so'
ERROR: /root/tensorflow/tensorflow/core/BUILD:1558:1: Target '//tensorflow/core/kernels:dataset_ops' contains an error and its package is in error and referenced by '//tensorflow/core:all_kernels'
ERROR: /root/tensorflow/tensorflow/core/BUILD:1558:1: Target '//tensorflow/core/kernels:list_kernels' contains an error and its package is in error and referenced by '//tensorflow/core:all_kernels'
ERROR: Analysis of target '//tensorflow/tools/pip_package:build_pip_package' failed; build aborted: Cannot compute config conditions
INFO: Elapsed time: 13.208s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (366 packages loaded, 10066 targets configured)
```
